<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Pyokit data-structures &mdash; Pyokit 0.2.0 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Pyokit 0.2.0 documentation" href="index.html" />
    <link rel="next" title="Developer documentation" href="developer.html" />
    <link rel="prev" title="File indexes" href="file_indexes.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="developer.html" title="Developer documentation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="file_indexes.html" title="File indexes"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Pyokit 0.2.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="pyokit-data-structures">
<h1>Pyokit data-structures<a class="headerlink" href="#pyokit-data-structures" title="Permalink to this headline">¶</a></h1>
<div class="section" id="genomic-intervals">
<h2>Genomic Intervals<a class="headerlink" href="#genomic-intervals" title="Permalink to this headline">¶</a></h2>
<p>Genomic intervals in pyokit are represented using the <cite>GenomicInterval</cite> class.
The basic information needed to define an interval is the chromosome it&#8217;s on,
and the start and end indices for the itnerval. We consider the end interval to
be exclusive (i.e. it&#8217;s not included in the interval). All intervals also have a
DNA strand, which defaults to the positive strand if not set. Additionally,
intervals can be given names and scores. The <cite>GenomicInterval</cite> class is
described below. For more information about using objects created from this
class, see  <a class="reference internal" href="genomicIntervals.html#genomicintervalssection"><span>Genomic Intervals</span></a>.</p>
<div class="section" id="the-genomicinterval-class">
<h3>The GenomicInterval class<a class="headerlink" href="#the-genomicinterval-class" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="pyokit.datastruct.genomicInterval.GenomicInterval">
<em class="property">class </em><code class="descclassname">pyokit.datastruct.genomicInterval.</code><code class="descname">GenomicInterval</code><span class="sig-paren">(</span><em>chrom</em>, <em>start</em>, <em>end</em>, <em>name=None</em>, <em>score=None</em>, <em>strand=None</em>, <em>scoreType=&lt;type 'int'&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.genomicInterval.GenomicInterval" title="Permalink to this definition">¶</a></dt>
<dd><p>Represents contiguous segment of a genome; inclusive of start, but not end.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>chrom</strong> &#8211; The chromosome this genomic interval is on.</li>
<li><strong>start</strong> &#8211; The start of this genomic interval (inclusive).</li>
<li><strong>end</strong> &#8211; The end of this genomic interval (exclusive).</li>
<li><strong>name</strong> &#8211; A name associated with this genomic interval.</li>
<li><strong>score</strong> &#8211; A score associated with this genomic interval.</li>
<li><strong>strand</strong> &#8211; The DNA strand (+ or -) that this genomic interval is on.</li>
<li><strong>scoreType</strong> &#8211; The type (e.g. int, float) of the score associated with
this genomic interval</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="pyokit.datastruct.genomicInterval.GenomicInterval.distance">
<code class="descname">distance</code><span class="sig-paren">(</span><em>e</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.genomicInterval.GenomicInterval.distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Distance between this interval and e &#8211; number of nucleotides.</p>
<p>We consider intervals that overlap to have a distance of 0 to each other.
The distance between two intervals on different chromosomes is considered
undefined, and causes an exception to be raised.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the distance from this GenomicInterval to e.</td>
</tr>
<tr class="field-even field"><th class="field-name">Parameters:</th><td class="field-body"><strong>e</strong> &#8211; the other genomic interval to find the distance to.</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Raises GenomicIntervalError:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">if self and e are on different chromosomes.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.genomicInterval.GenomicInterval.intersects">
<code class="descname">intersects</code><span class="sig-paren">(</span><em>e</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.genomicInterval.GenomicInterval.intersects" title="Permalink to this definition">¶</a></dt>
<dd><p>Check whether e intersects self.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">true if this elements intersects the element e</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.genomicInterval.GenomicInterval.isNegativeStrand">
<code class="descname">isNegativeStrand</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.genomicInterval.GenomicInterval.isNegativeStrand" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if this genomic interval is on the negative strand.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if this element is on the negative strand</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.genomicInterval.GenomicInterval.isPositiveStrand">
<code class="descname">isPositiveStrand</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.genomicInterval.GenomicInterval.isPositiveStrand" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if this genomic region is on the positive strand.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if this element is on the positive strand</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.genomicInterval.GenomicInterval.sameRegion">
<code class="descname">sameRegion</code><span class="sig-paren">(</span><em>e</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.genomicInterval.GenomicInterval.sameRegion" title="Permalink to this definition">¶</a></dt>
<dd><p>Check whether self represents the same DNA region as e.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>e</strong> &#8211; genomic region to compare against</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">True if self and e are for the same region (ignores differences
in non-region related fields, such as name or score &#8211; but does
consider strand)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.genomicInterval.GenomicInterval.signedDistance">
<code class="descname">signedDistance</code><span class="sig-paren">(</span><em>e</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.genomicInterval.GenomicInterval.signedDistance" title="Permalink to this definition">¶</a></dt>
<dd><p>Signed distance between this interval and e &#8211; number of nucleotides.</p>
<p>Get the signed distance from this genomic interval to another one. We
consider intervals that overlap to have a distance of 0 to each other. The
distance between two intervals on different chromosomes is considered
undefined, and causes an exception to be raised. If e comes earlier than
self, the distance will be negative.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the signed distance from this GenomicInterval to e.</td>
</tr>
<tr class="field-even field"><th class="field-name">Parameters:</th><td class="field-body"><strong>e</strong> &#8211; the other genomic interval to find the distance to.</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Raises GenomicIntervalError:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body">if self and e are on different chromosomes.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.genomicInterval.GenomicInterval.sizeOfOverlap">
<code class="descname">sizeOfOverlap</code><span class="sig-paren">(</span><em>e</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.genomicInterval.GenomicInterval.sizeOfOverlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the size of the overlap between self and e.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">the number of bases that are shared in common between self and e.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.genomicInterval.GenomicInterval.subtract">
<code class="descname">subtract</code><span class="sig-paren">(</span><em>es</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.genomicInterval.GenomicInterval.subtract" title="Permalink to this definition">¶</a></dt>
<dd><p>Subtract the BED elements in es from self.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>es</strong> &#8211; a list of BED elements (or anything with chrom, start, end)</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a list of BED elements which represent what is left of
self after the subtraction. This might be an empty list.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.genomicInterval.GenomicInterval.transform_center">
<code class="descname">transform_center</code><span class="sig-paren">(</span><em>size</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.genomicInterval.GenomicInterval.transform_center" title="Permalink to this definition">¶</a></dt>
<dd><p>Tranform self so it is centered on the same spot, but has new size.</p>
<p>If the region grows, the extra nucleotides will be distributed evenly to
the 5&#8217; and 3&#8217; side of the current region if possible. If the extra is odd,
the 3&#8217; side will get the extra one. Similarly, if the resize shrinks the
interval, bases will be removed from the 5&#8217; and 3&#8217; sides equally; if the
number to remove is odd, the extra one will be removed from the 3&#8217; side.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>size</strong> &#8211; size of the region after transformation.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="interval-trees">
<span id="intervaltreessection"></span><h2>Interval Trees<a class="headerlink" href="#interval-trees" title="Permalink to this headline">¶</a></h2>
<p>Interval trees are binary trees that allow random access lookup of intervals
that are intersected by a given point or interval.</p>
<div class="section" id="the-intervaltree-class">
<h3>The IntervalTree class<a class="headerlink" href="#the-intervaltree-class" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="pyokit.datastruct.intervalTree.IntervalTree">
<em class="property">class </em><code class="descclassname">pyokit.datastruct.intervalTree.</code><code class="descname">IntervalTree</code><span class="sig-paren">(</span><em>intervals</em>, <em>openEnded=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.intervalTree.IntervalTree" title="Permalink to this definition">¶</a></dt>
<dd><p>An interval tree is a binary tree that allows fast O(log(n)) lookup of
intervals that intersect a given point or interval.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>intervals</strong> &#8211; list of intervals, doesn&#8217;t need to be sorted in any way.
Can be any object, as long as they have &#8216;start&#8217; and &#8216;end&#8217;
attributes.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="pyokit.datastruct.intervalTree.IntervalTree.intersectingInterval">
<code class="descname">intersectingInterval</code><span class="sig-paren">(</span><em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.intervalTree.IntervalTree.intersectingInterval" title="Permalink to this definition">¶</a></dt>
<dd><p>given an interval, get intervals in the tree that are intersected.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>start</strong> &#8211; start of the intersecting interval</li>
<li><strong>end</strong> &#8211; end of the intersecting interval</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">the list of intersected intervals</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.intervalTree.IntervalTree.intersectingIntervalIterator">
<code class="descname">intersectingIntervalIterator</code><span class="sig-paren">(</span><em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.intervalTree.IntervalTree.intersectingIntervalIterator" title="Permalink to this definition">¶</a></dt>
<dd><p>Get an iterator which will iterate over those objects in the tree which
intersect the given interval - sorted in order of start index</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>start</strong> &#8211; find intervals in the tree that intersect an interval with
with this start index (inclusive)</li>
<li><strong>end</strong> &#8211; find intervals in the tree that intersect an interval with
with this end index (exclusive)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">an iterator that will yield intersected intervals</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.intervalTree.IntervalTree.intersectingPoint">
<code class="descname">intersectingPoint</code><span class="sig-paren">(</span><em>p</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.intervalTree.IntervalTree.intersectingPoint" title="Permalink to this definition">¶</a></dt>
<dd><p>given a point, get intervals in the tree that are intersected.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>p</strong> &#8211; intersection point</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the list of intersected intervals</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="sequences">
<h2>Sequences<a class="headerlink" href="#sequences" title="Permalink to this headline">¶</a></h2>
<p>Sequences in Pyokit are represented using the Sequence class, which wraps a
name and the actual sequence data, and provides a lot of the basic functionality
for manipulating sequences. Specializations exist for particular sequence
formats. At present, fasta and fastq are supported. The Sequence, FastaSequence
and FastqSequence classes are described below. For more information about
manipulating sequence data, see <a class="reference internal" href="sequences.html#sequencessection"><span>Biological Sequences</span></a></p>
<div class="section" id="the-sequence-class">
<span id="sequenceclasssection"></span><h3>The Sequence class<a class="headerlink" href="#the-sequence-class" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="pyokit.datastruct.sequence.Sequence">
<em class="property">class </em><code class="descclassname">pyokit.datastruct.sequence.</code><code class="descname">Sequence</code><span class="sig-paren">(</span><em>seqName</em>, <em>seqData</em>, <em>start_coord=None</em>, <em>end_coord=None</em>, <em>strand='+'</em>, <em>remaining=0</em>, <em>meta_data=None</em>, <em>useMutableString=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>This is the base class for all sequences in Pyokit. Objects from this class
will have only a sequence name and actual nucleotide sequence data.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>seqName</strong> &#8211; A name describing the sequence. Can be any string.</li>
<li><strong>seqData</strong> &#8211; The nucleotide sequence data. Can be DNA or RNA.
Note that there is no check to make sure the
sequence data is valid, that&#8217;s the responsibility
of the caller.</li>
<li><strong>start_coord</strong> &#8211; TODO</li>
<li><strong>end_coord</strong> &#8211; TODO</li>
<li><strong>strand</strong> &#8211; By default, this is +, but can also be set to -
to indicate that this sequence is a reverse
complement.</li>
<li><strong>remaining</strong> &#8211; the amount of sequence that comes after this; 0 if
this is the whole sequence. Alterntively, you
might think of this as the negative strand
coordinates of the end of this sequence.</li>
<li><strong>meta_data</strong> &#8211; dictionary containing meta-data key-value pairs</li>
<li><strong>useMutableString</strong> &#8211; Store the sequence data as a mutable string,
rather than a regular python string. This should
make editing operations must faster, but it comes
at the expense of less flexibility (e.g. the
object can not be used as a hash key because it
is mutable.)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.clip_end">
<code class="descname">clip_end</code><span class="sig-paren">(</span><em>seq</em>, <em>mm_score</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.clip_end" title="Permalink to this definition">¶</a></dt>
<dd><p>Clip a sequence from the end of this sequence &#8211; we assume the sequence
to be clipped will always begin somewhere in this sequence, but may not
be fully contained. If found, replaced with Ns.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>seq</strong> &#8211; sequence to be clipped</li>
<li><strong>mm_score</strong> &#8211; the number of matching bases needed to consider a hit,
mm_score = len(seq) would be 100% match</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.copy">
<code class="descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Copy constructor for Sequence objects.</p>
</dd></dl>

<dl class="attribute">
<dt id="pyokit.datastruct.sequence.Sequence.effective_len">
<code class="descname">effective_len</code><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.effective_len" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the length of the sequence if N&#8217;s are disregarded.</p>
</dd></dl>

<dl class="attribute">
<dt id="pyokit.datastruct.sequence.Sequence.end">
<code class="descname">end</code><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.end" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The coordinate of the end of this sequence; as with all other</td>
</tr>
</tbody>
</table>
<p>indexing of sequences in pyokit, sequences are not inclusive of their last
index. Computed just-in-time from the ungapped sequence length if it
wasn&#8217;t provided at construction time.</p>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.gapped_relative_subsequence">
<code class="descname">gapped_relative_subsequence</code><span class="sig-paren">(</span><em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.gapped_relative_subsequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a subsequence from this sequence using coordinates that are
relative to the start of the sequence (relative position 1) and the number
of nuceltodies in the sequence, including gaps. For example:</p>
<blockquote>
<div>46  &#8211;&gt; A&#8211;CTGC-TAGC-GATCGACT &lt;&#8211;  62      subsequence(2,7) == &#8211;CTG</div></blockquote>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.isDNA">
<code class="descname">isDNA</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.isDNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a guess as to whether this sequence is a DNA sequence or not by
looking at the symbols it contains.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if contains only DNA nucleotides, False otherwise</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.isLowQuality">
<code class="descname">isLowQuality</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.isLowQuality" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine whether this is a low quality sequence. To be considered a low
quality sequence, it must have &gt; 10% Ns.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if this sequence meets the above definition of low-quality.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.isPolyA">
<code class="descname">isPolyA</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.isPolyA" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine whether this sequence is polyA. To be a polyA sequence, it
must have &gt; 90% Adenine.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if the sequence is PolyA by the above definition.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.isPolyT">
<code class="descname">isPolyT</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.isPolyT" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine whether this sequence is polyT. To be a polyT sequence, it
must have &gt; 90% Thymine.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if the sequence is PolyT by the above definition.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.isRNA">
<code class="descname">isRNA</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.isRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Make a guess as to whether this sequence is an RNA sequence or not by
looking at the symbols it contains.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">True if contains only RNA nucleotides, False otherwise</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.is_positive_strand">
<code class="descname">is_positive_strand</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.is_positive_strand" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.maskMatch">
<code class="descname">maskMatch</code><span class="sig-paren">(</span><em>mask</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.maskMatch" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine whether this sequence matches the given mask.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>mask</strong> &#8211; string to match against. Ns in the mask are considered to
match anything in the sequence &#8211; all other chars must
match exactly.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">True if the mask matches at all places, otherwise false</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.maskRegion">
<code class="descname">maskRegion</code><span class="sig-paren">(</span><em>region</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.maskRegion" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace nucleotides in this sequence in the regions given by Ns</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>region</strong> &#8211; any object with .start and .end attributes. Co-ords are
zero based and inclusive of both end points. Any other
attributes (e.g. chrom.) are ignored.</td>
</tr>
<tr class="field-even field"><th class="field-name" colspan="2">Raises SequenceError:</th></tr>
<tr class="field-even field"><td>&nbsp;</td><td class="field-body">if region specifies nucleotides not present in
this sequence</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.maskRegions">
<code class="descname">maskRegions</code><span class="sig-paren">(</span><em>regions</em>, <em>verbose=False</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.maskRegions" title="Permalink to this definition">¶</a></dt>
<dd><p>Mask the given regions in this sequence with Ns.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>region</strong> &#8211; iterable of regions to mask. Each region can be any object
with .start and .end attributes. Co-ords are zero based
and inclusive of both end points. Any other attributes
(e.g. chrom.) are ignored.</li>
<li><strong>verbose</strong> &#8211; print status messages to stderr if True</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.meta_data_to_string">
<code class="descname">meta_data_to_string</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.meta_data_to_string" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.nsLeft">
<code class="descname">nsLeft</code><span class="sig-paren">(</span><em>amount</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.nsLeft" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace leftmost &lt;amount&gt; bases by Ns.</p>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.nsRight">
<code class="descname">nsRight</code><span class="sig-paren">(</span><em>amount</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.nsRight" title="Permalink to this definition">¶</a></dt>
<dd><p>Replace rightmost &lt;amount&gt; bases by Ns</p>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.percentNuc">
<code class="descname">percentNuc</code><span class="sig-paren">(</span><em>nuc</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.percentNuc" title="Permalink to this definition">¶</a></dt>
<dd><p>return the percentage of the sequence which is equal to the passed nuc.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>nuc</strong> &#8211; the nucleotide to compute percentage composition for. There
is no check to make sure this is a valid nucleotide.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">the percentage of the sequence that is &lt;nuc&gt;</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.relative_subsequence">
<code class="descname">relative_subsequence</code><span class="sig-paren">(</span><em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.relative_subsequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a subsequence from this sequence using coordinates that are
relative to the start (relative position 1) and end coordinates of the
sequence. For example:</p>
<blockquote>
<div>46  &#8211;&gt; A&#8211;CTGC-TAGC-GATCGACT &lt;&#8211;  62      subsequence(2,7) == CTGC-T</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>start</strong> &#8211; the index marking the start (inclusive) of the
subsequence. This is a one-based index, and is in the
coordinate space of the sequence (i.e. from 1 to N,
where N is the number of non-gap nucleotides in the
sequence)</li>
<li><strong>end</strong> &#8211; the index marking the end (exclusive) of the
subsequence. This is a one-based index, and is in the
coordinate space of the sequence (i.e. from 1 to N,
where N is the number of non-gap nucleotides in the
sequence)</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">a new sequence object that represents the subsequence</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Rasie SequenceError:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body"><p class="first last">if the start coordinate is less than 1 or the end
coordinate is greater than the ungapped length of
this sequence.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.reverseComplement">
<code class="descname">reverseComplement</code><span class="sig-paren">(</span><em>isRNA=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.reverseComplement" title="Permalink to this definition">¶</a></dt>
<dd><p>Reverse complement this sequence in-place.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>isRNA</strong> &#8211; if True, treat this sequence as RNA. If False, treat it as
DNA. If None (default), inspect the sequence and make a
guess as to whether it is RNA or DNA.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.similarity">
<code class="descname">similarity</code><span class="sig-paren">(</span><em>self_start</em>, <em>self_end</em>, <em>other_start</em>, <em>other_end</em>, <em>other</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.similarity" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the number of matching bases in the subsequences self[start, end]
and other[o_start, o_end]. Note that the subsequences must be the same
length.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>self_start</strong> &#8211; start index for sub-sequence in self</li>
<li><strong>self_end</strong> &#8211; end index for sub-sequence in self</li>
<li><strong>other_start</strong> &#8211; start index for subsequence in other sequence</li>
<li><strong>other_end</strong> &#8211; end index for subsequence in other sequence</li>
<li><strong>other</strong> &#8211; other sequence to compare to this.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.split">
<code class="descname">split</code><span class="sig-paren">(</span><em>point=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.split" title="Permalink to this definition">¶</a></dt>
<dd><p>Split this sequence into two halves and return them. The original
sequence remains unmodified.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>point</strong> &#8211; defines the split point, if None then the centre is used</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">two Sequence objects &#8211; one for each side</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="pyokit.datastruct.sequence.Sequence.start">
<code class="descname">start</code><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.start" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">The coordinate of the first nucleotide in this sequence; by</td>
</tr>
</tbody>
</table>
<p>convention, we call this coordinate 1 if no other value was provided.</p>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.subsequence">
<code class="descname">subsequence</code><span class="sig-paren">(</span><em>start</em>, <em>end</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.subsequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract a subsequence from this sequence object using absolute coordinates
that exist in the same coordinate space as the sequence itself. For
example:</p>
<blockquote>
<div>46  &#8211;&gt; A&#8211;CTGC-TAGC-GATCGACT &lt;&#8211;  62      subsequence(47,52) == CTGC-T</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>start</strong> &#8211; the index marking the start (inclusive) of the
subsequence. This is a one-based index, and is in the
same coordinate space as this sequence object.</li>
<li><strong>end</strong> &#8211; the index marking the end (exclusive) of the
subsequence. This is a one-based index, and is in the
same coordinate space as this sequence object.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">a new sequence object that represents the subsequence of this from
position start (indexed from 1, inclusive) to end (indexed from
1, exclusive). Ungapped length of this will always be equal to
end - start</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name" colspan="2">Rasie SequenceError:</th></tr>
<tr class="field-odd field"><td>&nbsp;</td><td class="field-body"><p class="first last">if the coordinates given fall outside of the start
and end indices of this sequence object.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.toDNA">
<code class="descname">toDNA</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.toDNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert this sequence in-place to a DNA sequence by changing any Us to Ts</p>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.toRNA">
<code class="descname">toRNA</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.toRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert this sequence in-place to an RNA sequence by changing any Ts
to Us</p>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.to_fasta_str">
<code class="descname">to_fasta_str</code><span class="sig-paren">(</span><em>line_width=50</em>, <em>include_coords=True</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.to_fasta_str" title="Permalink to this definition">¶</a></dt>
<dd><table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">string representation of this sequence object in fasta format</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="pyokit.datastruct.sequence.Sequence.truncate">
<code class="descname">truncate</code><span class="sig-paren">(</span><em>newLength</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.sequence.Sequence.truncate" title="Permalink to this definition">¶</a></dt>
<dd><p>Truncate this sequence in-place so it&#8217;s only &lt;newLength&gt; nucleotides
long.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>newLength</strong> &#8211; the length to truncate this sequence to.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="the-fastasequence-class">
<span id="fastasequenceclasssection"></span><h3>The FastaSequence class<a class="headerlink" href="#the-fastasequence-class" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="the-fastqsequence-class">
<span id="fastqsequenceclasssection"></span><h3>The FastqSequence class<a class="headerlink" href="#the-fastqsequence-class" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="sequence-alignments">
<h2>Sequence Alignments<a class="headerlink" href="#sequence-alignments" title="Permalink to this headline">¶</a></h2>
<div class="section" id="the-pairwisealignment-class">
<span id="pairwisealignmentclasssection"></span><h3>The PairwiseAlignment class<a class="headerlink" href="#the-pairwisealignment-class" title="Permalink to this headline">¶</a></h3>
<dl class="class">
<dt id="pyokit.datastruct.multipleAlignment.PairwiseAlignment">
<em class="property">class </em><code class="descclassname">pyokit.datastruct.multipleAlignment.</code><code class="descname">PairwiseAlignment</code><span class="sig-paren">(</span><em>s1</em>, <em>s2</em>, <em>meta_data=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.multipleAlignment.PairwiseAlignment" title="Permalink to this definition">¶</a></dt>
<dd><p>An alignment of two sequences (DNA, RNA, protein...).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>s1</strong> &#8211; the first sequence, with gaps</li>
<li><strong>s2</strong> &#8211; the second sequence, with gaps</li>
<li><strong>meta_data</strong> &#8211; a dictionary with key-value pairs representing meta-data
about this alignment.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="pyokit.datastruct.multipleAlignment.PairwiseAlignment.s1">
<code class="descname">s1</code><a class="headerlink" href="#pyokit.datastruct.multipleAlignment.PairwiseAlignment.s1" title="Permalink to this definition">¶</a></dt>
<dd><p>:return the first sequences in the alignment.</p>
</dd></dl>

<dl class="attribute">
<dt id="pyokit.datastruct.multipleAlignment.PairwiseAlignment.s2">
<code class="descname">s2</code><a class="headerlink" href="#pyokit.datastruct.multipleAlignment.PairwiseAlignment.s2" title="Permalink to this definition">¶</a></dt>
<dd><p>:return the second sequences in the alignment.</p>
</dd></dl>

</dd></dl>

</div>
</div>
<div class="section" id="gene-ontology">
<span id="ontologyclasssection"></span><h2>Gene Ontology<a class="headerlink" href="#gene-ontology" title="Permalink to this headline">¶</a></h2>
<p>There are two classes for representing gene ontology data in Pyokit. The
first of these is a simple data type wrapping basic information about a GO
term:</p>
<dl class="class">
<dt id="pyokit.datastruct.geneOntology.GeneOntologyTerm">
<em class="property">class </em><code class="descclassname">pyokit.datastruct.geneOntology.</code><code class="descname">GeneOntologyTerm</code><span class="sig-paren">(</span><em>name</em>, <em>identified=None</em>, <em>catagory=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.geneOntology.GeneOntologyTerm" title="Permalink to this definition">¶</a></dt>
<dd><p>Represents a gene ontology term.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> &#8211; the term name, can contain spaces;
e.g.: intracellular transport</li>
<li><strong>identifier</strong> &#8211; the term identifier; e.g: <a class="reference external" href="GO:0046907">GO:0046907</a></li>
<li><strong>catagory</strong> &#8211; the database or catagory the term belongs to;
e.g.: GOTERM_BP_FAT</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<p>While the second is a subclass of the general GO term class which adds
additional information about term enrichment.</p>
<dl class="class">
<dt id="pyokit.datastruct.geneOntology.GeneOntologyEnrichmentResult">
<em class="property">class </em><code class="descclassname">pyokit.datastruct.geneOntology.</code><code class="descname">GeneOntologyEnrichmentResult</code><span class="sig-paren">(</span><em>name</em>, <em>pvalue</em>, <em>identifier=None</em>, <em>catagory=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pyokit.datastruct.geneOntology.GeneOntologyEnrichmentResult" title="Permalink to this definition">¶</a></dt>
<dd><p>Represents the result of a gene ontology enrichment calculation for a
single GO term.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> &#8211; the term name, can contain spaces;
e.g.: intracellular transport</li>
<li><strong>pvalue</strong> &#8211; significance of the enrichment of this term</li>
<li><strong>identifier</strong> &#8211; the term identifier; e.g: <a class="reference external" href="GO:0046907">GO:0046907</a></li>
<li><strong>catagory</strong> &#8211; the database or catagory the term belongs to;
e.g.: GOTERM_BP_FAT</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Pyokit data-structures</a><ul>
<li><a class="reference internal" href="#genomic-intervals">Genomic Intervals</a><ul>
<li><a class="reference internal" href="#the-genomicinterval-class">The GenomicInterval class</a></li>
</ul>
</li>
<li><a class="reference internal" href="#interval-trees">Interval Trees</a><ul>
<li><a class="reference internal" href="#the-intervaltree-class">The IntervalTree class</a></li>
</ul>
</li>
<li><a class="reference internal" href="#sequences">Sequences</a><ul>
<li><a class="reference internal" href="#the-sequence-class">The Sequence class</a></li>
<li><a class="reference internal" href="#the-fastasequence-class">The FastaSequence class</a></li>
<li><a class="reference internal" href="#the-fastqsequence-class">The FastqSequence class</a></li>
</ul>
</li>
<li><a class="reference internal" href="#sequence-alignments">Sequence Alignments</a><ul>
<li><a class="reference internal" href="#the-pairwisealignment-class">The PairwiseAlignment class</a></li>
</ul>
</li>
<li><a class="reference internal" href="#gene-ontology">Gene Ontology</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="file_indexes.html"
                        title="previous chapter">File indexes</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="developer.html"
                        title="next chapter">Developer documentation</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/datastrct.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="developer.html" title="Developer documentation"
             >next</a> |</li>
        <li class="right" >
          <a href="file_indexes.html" title="File indexes"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Pyokit 0.2.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2014, Philip J. Uren.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.3.
    </div>
  </body>
</html>