
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Index &mdash; Pyokit 0.2.0 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Pyokit 0.2.0 documentation" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="#" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="index.html">Pyokit 0.2.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            

<h1 id="index">Index</h1>

<div class="genindex-jumpbox">
 <a href="#_"><strong>_</strong></a>
 | <a href="#A"><strong>A</strong></a>
 | <a href="#B"><strong>B</strong></a>
 | <a href="#C"><strong>C</strong></a>
 | <a href="#D"><strong>D</strong></a>
 | <a href="#E"><strong>E</strong></a>
 | <a href="#F"><strong>F</strong></a>
 | <a href="#G"><strong>G</strong></a>
 | <a href="#H"><strong>H</strong></a>
 | <a href="#I"><strong>I</strong></a>
 | <a href="#K"><strong>K</strong></a>
 | <a href="#M"><strong>M</strong></a>
 | <a href="#N"><strong>N</strong></a>
 | <a href="#O"><strong>O</strong></a>
 | <a href="#P"><strong>P</strong></a>
 | <a href="#Q"><strong>Q</strong></a>
 | <a href="#R"><strong>R</strong></a>
 | <a href="#S"><strong>S</strong></a>
 | <a href="#T"><strong>T</strong></a>
 | <a href="#U"><strong>U</strong></a>
 
</div>
<h2 id="_">_</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="sequence_alignments.html#pyokit.io.alignmentIterators._rm_parse_header_line">_rm_parse_header_line() (in module pyokit.io.alignmentIterators)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="A">A</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="ui.html#pyokit.interface.cli.CLI.addOption">addOption() (pyokit.interface.cli.CLI method)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.ALIG_SCORE_KEY">ALIG_SCORE_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.multipleAlignment.PairwiseAlignment.alignment_to_sequence_coords">alignment_to_sequence_coords() (pyokit.datastruct.multipleAlignment.PairwiseAlignment method)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.ANNOTATION_KEY">ANNOTATION_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="B">B</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="bed.html#pyokit.io.bedIterators.BEDIterator">BEDIterator() (in module pyokit.io.bedIterators)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="genomicIntervals_manip.html#pyokit.datastruct.genomicInterval.bucketIterator">bucketIterator() (in module pyokit.datastruct.genomicInterval)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="C">C</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="ui.html#pyokit.interface.cli.CLI">CLI (class in pyokit.interface.cli)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.clipAdaptor">clipAdaptor() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.clipThreePrime">clipThreePrime() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="genomicIntervals_manip.html#pyokit.datastruct.genomicInterval.collapseRegions">collapseRegions() (in module pyokit.datastruct.genomicInterval)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.containsAdaptor">containsAdaptor() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.copy">copy() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="D">D</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="gene_ontology.html#pyokit.io.david.david_results_iterator">david_results_iterator() (in module pyokit.io.david)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.genomicInterval.GenomicInterval.distance">distance() (pyokit.datastruct.genomicInterval.GenomicInterval method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="E">E</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.effectiveLength">effectiveLength() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="F">F</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="sequences.html#pyokit.io.fastaIterators.fastaIterator">fastaIterator() (in module pyokit.io.fastaIterators)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.FastaSequence">FastaSequence (class in pyokit.datastruct.sequence)</a>
  </dt>

      
  <dt><a href="sequences.html#pyokit.io.fastqIterators.fastqIterator">fastqIterator() (in module pyokit.io.fastqIterators)</a>
  </dt>

      
  <dt><a href="sequences.html#pyokit.io.fastqIterators.fastqIteratorComplex">fastqIteratorComplex() (in module pyokit.io.fastqIterators)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="sequences.html#pyokit.io.fastqIterators.fastqIteratorSimple">fastqIteratorSimple() (in module pyokit.io.fastqIterators)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.FastqSequence">FastqSequence (class in pyokit.datastruct.sequence)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.FastaSequence.formattedString">formattedString() (pyokit.datastruct.sequence.FastaSequence method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="G">G</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.geneOntology.GeneOntologyEnrichmentResult">GeneOntologyEnrichmentResult (class in pyokit.datastruct.geneOntology)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.geneOntology.GeneOntologyTerm">GeneOntologyTerm (class in pyokit.datastruct.geneOntology)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.genomicInterval.GenomicInterval">GenomicInterval (class in pyokit.datastruct.genomicInterval)</a>
  </dt>

      
  <dt><a href="ui.html#pyokit.interface.cli.CLI.getAllArguments">getAllArguments() (pyokit.interface.cli.CLI method)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="ui.html#pyokit.interface.cli.CLI.getArgument">getArgument() (pyokit.interface.cli.CLI method)</a>
  </dt>

      
  <dt><a href="ui.html#pyokit.interface.cli.CLI.getOption">getOption() (pyokit.interface.cli.CLI method)</a>
  </dt>

      
  <dt><a href="ui.html#pyokit.interface.cli.CLI.getValue">getValue() (pyokit.interface.cli.CLI method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="H">H</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="ui.html#pyokit.interface.cli.CLI.hasArgument">hasArgument() (pyokit.interface.cli.CLI method)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="ui.html#pyokit.interface.cli.CLI.hasOption">hasOption() (pyokit.interface.cli.CLI method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="I">I</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.intervalTree.IntervalTree.intersectingInterval">intersectingInterval() (pyokit.datastruct.intervalTree.IntervalTree method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.intervalTree.IntervalTree.intersectingIntervalIterator">intersectingIntervalIterator() (pyokit.datastruct.intervalTree.IntervalTree method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.intervalTree.IntervalTree.intersectingPoint">intersectingPoint() (pyokit.datastruct.intervalTree.IntervalTree method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.genomicInterval.GenomicInterval.intersects">intersects() (pyokit.datastruct.genomicInterval.GenomicInterval method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.intervalTree.IntervalTree">IntervalTree (class in pyokit.datastruct.intervalTree)</a>
  </dt>

      
  <dt><a href="bed.html#pyokit.io.bedIterators.intervalTrees">intervalTrees() (in module pyokit.io.bedIterators)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.isDNA">isDNA() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.isLowQuality">isLowQuality() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.genomicInterval.GenomicInterval.isNegativeStrand">isNegativeStrand() (pyokit.datastruct.genomicInterval.GenomicInterval method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.isPolyA">isPolyA() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.isPolyT">isPolyT() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.genomicInterval.GenomicInterval.isPositiveStrand">isPositiveStrand() (pyokit.datastruct.genomicInterval.GenomicInterval method)</a>
  </dt>

      
  <dt><a href="ui.html#pyokit.interface.cli.Option.isRequired">isRequired() (pyokit.interface.cli.Option method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.isRNA">isRNA() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      
  <dt><a href="ui.html#pyokit.interface.cli.Option.isSet">isSet() (pyokit.interface.cli.Option method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="K">K</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.KNOWN_KEYS">KNOWN_KEYS (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="M">M</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.maskMatch">maskMatch() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.maskRegion">maskRegion() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.maskRegions">maskRegions() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.FastqSequence.merge">merge() (pyokit.datastruct.sequence.FastqSequence method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="N">N</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.nsLeft">nsLeft() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.nsRight">nsRight() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="O">O</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="ui.html#pyokit.interface.cli.Option">Option (class in pyokit.interface.cli)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="ui.html#pyokit.interface.cli.CLI.optionIsSet">optionIsSet() (pyokit.interface.cli.CLI method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="P">P</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="bed.html#pyokit.io.bedIterators.pairedBEDIterator">pairedBEDIterator() (in module pyokit.io.bedIterators)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.multipleAlignment.PairwiseAlignment">PairwiseAlignment (class in pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="ui.html#pyokit.interface.cli.CLI.parseCommandLine">parseCommandLine() (pyokit.interface.cli.CLI method)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.PCENT_S1_INDELS_KEY">PCENT_S1_INDELS_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.PCENT_S2_INDELS_KEY">PCENT_S2_INDELS_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.percentNuc">percentNuc() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="Q">Q</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.FastqSequence.qualityToSolexa">qualityToSolexa() (pyokit.datastruct.sequence.FastqSequence method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="R">R</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="genomicIntervals_manip.html#pyokit.datastruct.genomicInterval.regionsIntersection">regionsIntersection() (in module pyokit.datastruct.genomicInterval)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.io.alignmentIterators.repeat_masker_alignment_iterator">repeat_masker_alignment_iterator() (in module pyokit.io.alignmentIterators)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.multipleAlignment.PairwiseAlignment.repeat_masker_header">repeat_masker_header() (pyokit.datastruct.multipleAlignment.PairwiseAlignment method)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.FastqSequence.reverseComplement">reverseComplement() (pyokit.datastruct.sequence.FastqSequence method)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.reverseComplement">(pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.RM_ID_KEY">RM_ID_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="S">S</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.S1_END_KEY">S1_END_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.S1_END_NEG_STRAND_KEY">S1_END_NEG_STRAND_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.multipleAlignment.PairwiseAlignment.s1_is_reverse_comp">s1_is_reverse_comp() (pyokit.datastruct.multipleAlignment.PairwiseAlignment method)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.S1_NAME_KEY">S1_NAME_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.S1_START_KEY">S1_START_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.S1_START_NEG_STRAND_KEY">S1_START_NEG_STRAND_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.multipleAlignment.PairwiseAlignment.s1_ungapped_len">s1_ungapped_len (pyokit.datastruct.multipleAlignment.PairwiseAlignment attribute)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.S2_END_KEY">S2_END_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.S2_END_NEG_STRAND_KEY">S2_END_NEG_STRAND_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.multipleAlignment.PairwiseAlignment.s2_is_reverse_comp">s2_is_reverse_comp() (pyokit.datastruct.multipleAlignment.PairwiseAlignment method)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.S2_NAME_KEY">S2_NAME_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.S2_START_KEY">S2_START_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.S2_START_NEG_STRAND_KEY">S2_START_NEG_STRAND_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.multipleAlignment.PairwiseAlignment.s2_ungapped_len">s2_ungapped_len (pyokit.datastruct.multipleAlignment.PairwiseAlignment attribute)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.genomicInterval.GenomicInterval.sameRegion">sameRegion() (pyokit.datastruct.genomicInterval.GenomicInterval method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence">Sequence (class in pyokit.datastruct.sequence)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.multipleAlignment.PairwiseAlignment.sequence_to_alignment_coords">sequence_to_alignment_coords() (pyokit.datastruct.multipleAlignment.PairwiseAlignment method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.genomicInterval.GenomicInterval.signedDistance">signedDistance() (pyokit.datastruct.genomicInterval.GenomicInterval method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.similarity">similarity() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.multipleAlignment.PairwiseAlignment.size">size() (pyokit.datastruct.multipleAlignment.PairwiseAlignment method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.genomicInterval.GenomicInterval.sizeOfOverlap">sizeOfOverlap() (pyokit.datastruct.genomicInterval.GenomicInterval method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.FastqSequence.split">split() (pyokit.datastruct.sequence.FastqSequence method)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.split">(pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      </dl></dd>
      
  <dt><a href="datastrct.html#pyokit.datastruct.genomicInterval.GenomicInterval.subtract">subtract() (pyokit.datastruct.genomicInterval.GenomicInterval method)</a>
  </dt>

  </dl></td>
</tr></table>

<h2 id="T">T</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.multipleAlignment.PairwiseAlignment.to_repeat_masker_string">to_repeat_masker_string() (pyokit.datastruct.multipleAlignment.PairwiseAlignment method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.toDNA">toDNA() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.toRNA">toRNA() (pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.FastqSequence.trimLeft">trimLeft() (pyokit.datastruct.sequence.FastqSequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.FastqSequence.trimRight">trimRight() (pyokit.datastruct.sequence.FastqSequence method)</a>
  </dt>

      
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.FastqSequence.truncate">truncate() (pyokit.datastruct.sequence.FastqSequence method)</a>
  </dt>

      <dd><dl>
        
  <dt><a href="datastrct.html#pyokit.datastruct.sequence.Sequence.truncate">(pyokit.datastruct.sequence.Sequence method)</a>
  </dt>

      </dl></dd>
  </dl></td>
</tr></table>

<h2 id="U">U</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="sequence_alignments.html#pyokit.datastruct.multipleAlignment.UNKNOWN_RM_HEADER_FIELD_KEY">UNKNOWN_RM_HEADER_FIELD_KEY (in module pyokit.datastruct.multipleAlignment)</a>
  </dt>

  </dl></td>
  <td style="width: 33%" valign="top"><dl>
      
  <dt><a href="ui.html#pyokit.interface.cli.CLI.usage">usage() (pyokit.interface.cli.CLI method)</a>
  </dt>

  </dl></td>
</tr></table>



          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">

   

<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="#" title="General Index"
             >index</a></li>
        <li><a href="index.html">Pyokit 0.2.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Philip J. Uren.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
  </body>
</html>